controllers/common/lock.go

```diff
@@ -0,0 +1,44 @@
+// Copyright Contributors to the Open Cluster Management project
+
+package common
+
+import (
+	"sync"
+
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// PoliciesLock allows locking at the root policy level so that multiple controllers/goroutines don't race with each
+// other and cause unnecessary reconcile retries.
+type PoliciesLock struct {
+	policies     map[types.NamespacedName]*sync.Mutex
+	internalLock sync.Mutex
+}
+
+func (p *PoliciesLock) Lock(policy types.NamespacedName) {
+	p.internalLock.Lock()
+
+	if p.policies == nil {
+		p.policies = map[types.NamespacedName]*sync.Mutex{}
+	}
+
+	if _, ok := p.policies[policy]; !ok {
+		p.policies[policy] = &sync.Mutex{}
+	}
+
+	lock := p.policies[policy]
+
+	p.internalLock.Unlock()
+
+	lock.Lock()
```

Comment:
`lock.Lock()` is blocking if the lock is already held for that root policy, so the internal lock is released after reading from the map is done to avoid having an artificial lock on all root policies.

Reply:
I think that makes sense. But now I'm wondering if https://pkg.go.dev/sync#Map can be used here?